Documentation/connectors/oauth.md

```diff
@@ -0,0 +1,49 @@
+# Authentication using Generic OAuth 2.0 provider
+
+## Overview
+
+Dex users can make use of this connector to work with standards-compliant [OAuth 2.0](https://oauth.net/2/) authorization provider, in case of that authorization provider is not in the Dex connectors list.
+
+## Configuration
+
+The following is an example of a configuration for using OAuth connector with Reddit.
+
+```yaml
+connectors:
+- type: oauth
+  # ID of OAuth 2.0 provider
+  id: reddit 
+  # Name of OAuth 2.0 provider
+  name: reddit
+  config:
+    # Connector config values starting with a "$" will read from the environment.
+    clientID: $REDDIT_CLIENT_ID
+    clientSecret: $REDDIT_CLIENT_SECRET
+    redirectURI: http://127.0.0.1:5556/callback
+
+    tokenURL: https://www.reddit.com/api/v1/access_token
+    authorizationURL: https://www.reddit.com/api/v1/authorize
+    userInfoURL: https: https://www.reddit.com/api/v1/me
+ 
+    # Optional: Specify whether to communicate to Auth provider without validating SSL certificates
+    # insecureSkipVerify: false
+
+    # Optional: The location of file containing SSL certificates to commmunicate to Auth provider 
+    # rootCAs: /etc/ssl/reddit.pem
+
+    # Optional: List of scopes to request Auth provider for access user account
+    # scopes:
+    #  - identity
+
+    # Optional: Configurable keys for groups claim look up
+    # Default: groups
+    # groupsKey:
+
+    # Optional: Configurable keys for user ID claim look up
+    # Default: user_id
+    # userIDKey:
+
+    # Optional: Configurable keys for preferred username claim look up
+    # Default: preferred_username
+    # preferredUsernameKey:
```

Comment:
Can we add a claim mapping for these (except userIDKey) similar to the one introduced in #1634?

(`userIDKey` is not actually an OIDC claim, so it should remain separate)

Reply:
I wonder if `userIDKey` is not a standard claim (actullay i believe for oauth framework there is no standard claim, compared to OIDC), should we also include it in the claim mapping so user can configure it to match whatever they want from the auth provider?

I think the same case also apply to `email` and `email_verified`. Please correct me if there is a doc about standard claims of oauth.

Reply:
I think in our case claims are interpreted from the Dex providing OIDC point of view. No, they are not oauth claims, but we map some information from oauth to the OIDC token.

The userIDKey is only used by Dex to determine which field to use as the user identifier for the specific connector. The rest of the keys are "mapped" into an OIDC token as standard claims.

That's how I see it at least, correct me if my thought process is wrong.

Reply:
I have update the PR to include all identities mapping in a `ClaimMappings`. That is for all different kinds of non-standard user identity profile return by Oauth2 provider. I will list the user profile response from Reddit, AWS and LinkedIn for example.

[Reddit](https://github.com/reddit-archive/reddit/wiki/OAuth2-Quick-Start-Example#curl-example)
```
https://oauth.reddit.com/api/v1/me
{
    "comment_karma": 0, 
    "created": 1389649907.0, 
    "created_utc": 1389649907.0, 
    "has_mail": false,    
    "has_mod_mail": false, 
    "has_verified_email": null,   <--- emailVerifiedKey
    "id": "1",                               <--- userIDKey
    "is_gold": false, 
    "is_mod": true, 
    "link_karma": 1, 
    "name": "reddit_bot",           <--- preferredUsernameKey
    "over_18": true
}
```

[AWS](https://docs.aws.amazon.com/cognito/latest/developerguide/userinfo-endpoint.html)
```
/oauth2/userInfo
{
   "sub": "248289761001",      <--- userIDKey
   "name": "Jane Doe",           <--- userNameKey
   "given_name": "Jane",
   "family_name": "Doe",
   "preferred_username": "j.doe", <--- preferredUsernameKey
   "email": "janedoe@example.com"  <--- emailKey
}
```

[LinkedIn](https://docs.microsoft.com/en-us/linkedin/shared/integrations/people/profile-api)
```
   ...
   "localizedFirstName": "Bob",      <--- userNameKey
   "localizedHeadline": "API Enthusiast at LinkedIn",
   "vanityName": "bsmith",             <--- preferredUsernameKey
   "id":"yrZCpj2Z12",                      <--- userIDKey
   "localizedLastName": "Smith",
  ...
```
As you can see, the field for unique user identity could be `sub` or `id`, and the username field could has different names `vanityName` or `name`. So does the email verified key.

Id prefer this Oauth connector to be as general as possible so it could retrieve user profile from all kinds of responses from different provider and map them to dex identity.

Reply:
I have put eveything but `userIDKey` under `ClaimMappings` and added an error if user ID is not found since it is critical. `EmailKey` and `EmailVerifiedKey` are added so it can handle a response like reddit:

[Reddit](https://github.com/reddit-archive/reddit/wiki/OAuth2-Quick-Start-Example#curl-example)
```
https://oauth.reddit.com/api/v1/me
{
    "comment_karma": 0, 
    "created": 1389649907.0, 
    "created_utc": 1389649907.0, 
    "has_mail": false,    
    "has_mod_mail": false, 
    "has_verified_email": null,   <--- emailVerifiedKey
    "id": "1",                               <--- userIDKey
    "is_gold": false, 
    "is_mod": true, 
    "link_karma": 1, 
    "name": "reddit_bot",           <--- preferredUsernameKey
    "over_18": true
}
```